### PR TITLE
Decouple email and service name

### PR DESCRIPTION
**Pull request by dorianjp**
https://trello.com/c/yeyTL5es/158-decouple-email-and-service-name

Added 2 endpoints for validating service name and email from independently.
* /name/unique
* /email-from/unique

Left the old endpoint /unique for migration purposes, will be removed once admin is merged with usage of new endpoints.
* /unique

### CODE REVIEW

**Review by AntoineAugusti — CHANGES_REQUESTED** [review]

File: app/service/rest.py [hunk 1]
```diff
@@ -867,6 +867,29 @@ def is_service_name_unique():
...
+
+
+@service_blueprint.route('/email-from/unique', methods=["GET"])
+def is_service_email_from_unique():
+    service_id, email_from = check_unique_email_from_request_args(request)
+
+    email_from_exists = Service.query.filter(
+        Service.email_from == email_from,
+        Service.id != service_id
+    ).first()
```
[the remark this answers is not in the document]
Comment (dorianjp) [reply app/service/rest.py:887]: Maybe I'm misunderstanding, but a constraint in what context?

[the remark this answers is not in the document]
Comment (AntoineAugusti) [reply app/service/rest.py:887]: A unique index on `email_from` by service ID

File: app/service/rest.py [hunk 2]
```diff
@@ -980,7 +1003,30 @@ def create_smtp_relay(service_id):
...
+    service_id = request.args.get('service_id')
+    email_from = request.args.get('email_from', None)
+    errors = []
+    if not service_id:
+        errors.append({'service_id': ["Can't be empty"]})
+    if not email_from:
+        errors.append({'email_from': ["Can't be empty"]})
+    if errors:
+        raise InvalidRequest(errors, status_code=400)
+    return service_id, email_from
```
[the remark this answers is not in the document]
Comment (dorianjp) [reply app/service/rest.py:1026]: As discussed, schema validation will not be required with this type of GET endpoint due to impacts to data structures, test regressions, admin changes.

File: app/service/rest.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dorianjp) [reply app/service/rest.py:887]: uniqueness is already present, no index required at this time

**Review by dorianjp — COMMENTED** [review]
will stand as discussed, thanks @AntoineAugusti

File: app/service/rest.py [hunk 3]
```diff
@@ -980,7 +1003,30 @@ def create_smtp_relay(service_id):
...
 
-def check_request_args(request):
+def check_unique_name_request_args(request):
+    name = request.args.get('name', None)
+    errors = []
+    if not name:
+        errors.append({'name': ["Can't be empty"]})
+    if errors:
+        raise InvalidRequest(errors, status_code=400)
+    return name
```
Comment (AntoineAugusti) [inline app/service/rest.py:1013]:
```suggestion
    if not name:
        raise InvalidRequest({'name': ["Can't be empty"]}, status_code=400)
    return name
```

If you want to simplify only

**Review by AntoineAugusti — APPROVED** [review]
LGTM!

File: app/service/rest.py (hunk 3 above)
Comment (dorianjp) [reply app/service/rest.py]: I believed in you and you broke my tests!

Always double check 'fast & risky' @AntoineAugusti is my new motto.

Comment (AntoineAugusti) [reply app/service/rest.py]: Sorry about that!

**Comment by dorianjp** [comment]
[Decouple Email and Service Name](https://trello.com/c/yeyTL5es/158-decouple-email-and-service-name)

**Pull request merged**